builders/build/substrate/adding-custom-made-module.md

```diff
@@ -0,0 +1,338 @@
+        // 3. Checks that the user is not already participating
+        // 4. Adds the user as a new participant for the prize
+        // 5. Transfers the ticket cost to the module's account, to be hold until transferred to the winner
+        // 6. Notify the event
+    
+    }
+
+    #[pallet::call_index(1)]
+    #[pallet::weight(0)]
+    pub fn award_prize(origin: OriginFor<T>) -> DispatchResult {
+
+        // 1. Validates the origin signature
+        // 2. Gets a random number from the randomness module
+        // 3. Selects the winner from the participants lit
+        // 4. Transfers the total prize to the winner's account
+        // 5. Resets the participants list, and gets ready for another lottery round
+
+    }
+}
+```
+
+These calls also emit events, to keep the user informed and can return errors, should any of the validations go wrong.
+
+Here is the complete implementation of the calls, with the custom lottery logic:
+
+??? code "View the complete calls code"
+
+    ```rust
+    --8<-- 'code/basic-substrate/lottery-example-calls.rs'
+    ```
+
+### Implementing Custom Errors {: #implementing-custom-errors}
+
+The `#[pallet::error]` macro is applied to an enumeration of errors that might occur during the execution. It is important for security reasons to handle all error cases gracefully and never crash in the runtime.
+
+```rust
+// Errors inform users that something went wrong.
+#[pallet::error]
+pub enum Error<T> {
+    NotEnoughCurrency,
+    AccountAlreadyParticipating,
+    CanNotAddParticipant,
+}
+```
+
+### Implementing Events {: #implementing-events }
+
+The `#[pallet::event]` macro is applied to an enumeration of events to inform the user of any changes in the state or important actions that happened during the execution in the runtime.
+
+The following snippet shows the events defined in the lottery example:
+
+```rust
+#[pallet::event]
+#[pallet::generate_deposit(pub(super) fn deposit_event)]
+pub enum Event<T: Config> {
+    /// Event emitted when a ticket is bought
+    TicketBought { who: T::AccountId },
+    /// Event emitted when the prize is awarded
+    PrizeAwarded { winner: T::AccountId },
+    /// Event emitted when there are no participants
+    ThereAreNoParticipants,
+}
+```
+
+### Implementing Storage for State Persistence {: #implementing-storage }
+
+The `#[pallet::storage]` macro initializes a runtime storage structure. In this example, a basic value storage structure is used to persist the list of participants in a bounded capacity vector ([BoundedVec](https://crates.parity.io/frame_support/storage/bounded_vec/struct.BoundedVec.html){target=_blank}).
+
+In the heavily constrained environment of an AppChain, deciding what to store and which structure to use can be critical in terms of performance. More on this topic is covered in the [Substrate documentation](https://docs.substrate.io/build/runtime-storage/){target=_blank}.
```

Comment:
If you agree with my suggestion above, then delete this paragraph

Reply:
Agreed!